_posts/2023-08-25-using-kani-to-validate-security-boundaries-in-aws-firecracker.md

```diff
@@ -0,0 +1,292 @@
+
+The property we want to verify is that a microVM is not allowed to exceed the configured maximum I/O throughput rate. For a virtual block device rate-limited at 1GB/s, we want to prove that in any one-second interval, at most 1GB of data is allowed to pass through the device. 
+
+What sounds simple in theory is actually fairly difficult to implement. For example, due to a [rounding error](https://github.com/firecracker-microvm/firecracker/pull/3706) a guest could, in some scenarios, do up to 0.01% more I/O than configured. We discovered this bug thanks to a Kani harness for our throughput property stated above, and this harnesses is the main focus of the rest of this section.
+
+### Teaching Kani about Time
+
+The core component of our rate-limiting implementation is a `TokenBucket`.  In Firecracker, we define it as
+
+```rs
+pub struct TokenBucket {
+    // Maximal number of tokens this bucket can hold.
+    size: u64,
+
+     // Complete refill time in milliseconds.
+    refill_time: u64,
+
+    // Current token budget.
+    budget: u64,
+    
+    // Last time this token bucket was replenished.
+    last_update: Instant,
+
+    // -- snip --
+}
+```
+
+It offers an `auto_replenish` function which computes how many tokens the leaky bucket algorithm should have generated since `last_update` (and then updates `last_update` accordingly).
+
+A `TokenBucket` is inherently tied to time-related APIs such as `std::time::Instant`, for which Kani does not have built-in support. This means it is not able to reason about `TokenBucket`s. To solve this problem, we use Kani’s [stubbing](https://model-checking.github.io/kani-verifier-blog/2023/02/28/kani-internship-projects-2022-stubbing.html) to provide a model for the `Instant::now` function. Since Firecracker uses a monotonic clock for its rate-limiting, this stub needs to return non-deterministic monotonically increasing instants.
+
+However, when trying to stub `now`, one will quickly notice that `Instant` does not offer any constructors for creating an instance from, say, a Unix timestamp. In fact, it is impossible to construct an `Instant` outside of the standard library as its fields are private. When in such a situation, the solution is often to go down the call stack of the function that you want to stub, to see if any of the functions further down can be stubbed out instead to achieve the desired effect. In our case, `now` calls functions in (private) OS specific time modules, until it bottoms out at [`libc::clock_gettime`](https://www.gnu.org/software/libc/manual/html_node/Getting-the-Time.html#index-clock_005fgettime).
+
+The `clock_gettime` function is passed a pointer to a `libc::timespec` structure, and the `tv_sec` and `tv_nsec` members of this structure are later used to construct the `Instant` returned by `Instant::now`. Therefore, we can use the following stub to achieve our goal of getting non-deterministic, monotonically non-decreasing `Instant`s:
```

Comment:
```suggestion
The `clock_gettime` function is passed a pointer to a `libc::timespec` structure, and the `tv_sec` and `tv_nsec` members of this structure are later used to construct the `Instant` returned by `Instant::now`. Therefore, we can use the following stub to achieve our goal of getting non-deterministic, monotonically increasing `Instant`s:
```

Reply:
To remain consistent with the previous text.

Reply:
Yup, need to be consistent, but "non-increasing" is correct. I replaced the other instance of "increasing" instead.